### ISSUE TITLE
UPX 3.95: "Cross-compressing" busybox-armv7l produce unusable executable

### ISSUE THREAD

**Issue by UMRnInside** [upx/upx#325]
## What's the problem (or question)?
On Debian 10 "buster" x86_64 (with `qemu-uset-static` and `upx-ucl` ):
```
wget https://busybox.net/downloads/binaries/1.28.1-defconfig-multiarch/busybox-armv7l
chmod a+x busybox-armv7l
```
Then:
```
user@host:/tmp$ upx busybox-armv7l 
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2018
UPX 3.95        Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 26th 2018

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   1079156 ->    622660   57.70%    linux/arm    busybox-armv7l                

Packed 1 file.
user@host:/tmp$ ./busybox-armv7l 
./busybox-armv7l: Invalid argument
```

The compressed executable reported `Invalid argument`

## What should have happened?
The compressed `busybox-armv7l` should run as the uncompressed one.

## Do you have an idea for a solution?
* I have no idea :disappointed: 
* After uncompressed this, the busybox-armv7l worked (w/o errors):
```
user@host:/tmp$ upx -d busybox-armv7l 
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2018
UPX 3.95        Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 26th 2018

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   1079156 <-    622660   57.70%    linux/arm    busybox-armv7l

Unpacked 1 file.
user@host:/tmp$ ./busybox-armv7l 
BusyBox v1.28.1 (2018-02-15 14:34:02 CET) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.
```
* On Debian 10 arm64, It works perfectly.
* `master` and `devel` branch of `upx` does *NOT* work, either (with libucl-dev 1.03+repack-5)
* Not sure if it's related to `qemu-user-static`

## How can we reproduce the issue?

1. Get a computer running Debian 10 "buster" **x86_64**
2. `apt install qemu-user-static upx-ucl`
3. `wget https://busybox.net/downloads/binaries/1.28.1-defconfig-multiarch/busybox-armv7l && chmod a+x busybox-armv7l`
4. `upx busybox-armv7l`
5. `./busybox-armv7l`

## Please tell us details about your environment.

* UPX version used (`upx --version`): `upx 3.95`
* Host Operating System and version: `Debian 10 "buster"`
* Host CPU architecture: `x86_64`
* Target Operating System and version: `Debian 10 "buster"` and "OpenWRT 18.06"
* Target CPU architecture: `arm`

**Comment by jreiser** [comment]
```
$ wget https://busybox.net/downloads/binaries/1.28.1-defconfig-multiarch/busybox-armv7l
$ file busybox-armv7l
busybox-armv7l: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, stripped
$ chmod +x ./busybox-armv7l
$ uname -m
x86_64
$ ./busybox-armv7l  # exactly step 5 of the recipe to reproduce
bash: ./busybox-armv7l: cannot execute binary file: Exec format error
```
so `busybox-armv7l` runs only on armv7l, and not on x86_64.  Compressing with UPX does not change that.  UPX detects the intended hardware architecture using the ELF header information of each input file, then compresses accordingly.  UPX does not change the hardware architecture.

**Comment by jreiser** [comment]
If the compressed `busybox-armv7l` (called `foo` below) is run on an ARM 32-bit machine, then I see correct output, such as this run under strace on Fedora 31:
```
$ uname -a
Linux f31arm32.domain 5.3.12-300.fc31.armv7hl #1 SMP Fri Nov 22 07:28:58 UTC 2019 armv7l armv7l armv7l GNU/Linux
$ strace ./foo
execve("./foo", ["./foo"], 0xbef804f0 /* 28 vars */) = 0
open("/proc/self/exe", O_RDONLY)        = 3
mmap2(NULL, 623204, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xb6ef7000
mmap2(0xb6ef7000, 622332, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0) = 0xb6ef7000
cacheflush(0xb6f8e7ec, 0xb6f8f264, 0)   = 0
mprotect(0xb6f8e000, 4708, PROT_READ|PROT_EXEC) = 0
readlink("/proc/self/exe", "/home/user01/foo", 4095) = 17
cacheflush(0xbea83acc, 0xbea83b80, 0)   = 0
mmap2(0x10000, 1150976, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x10000
mmap2(0x10000, 1077908, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x10000
cacheflush(0x10000, 0x100b4, 0)         = 0
cacheflush(0x100b4, 0x117294, 0)        = 0
cacheflush(0x117294, 0x11729c, 0)       = 0
mprotect(0x10000, 1077908, PROT_READ|PROT_EXEC) = 0
mmap2(0x127000, 1181, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x127000
cacheflush(0x127294, 0x12749d, 0)       = 0
mprotect(0x127000, 1181, PROT_READ|PROT_WRITE) = 0
mmap2(0x128000, 3196, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x128000
brk(0x129000)                           = 0x203000
mmap2(NULL, 4096, PROT_READ, MAP_PRIVATE, 3, 0) = 0xb6ef6000
close(3)                                = 0
munmap(0xb6ef7000, 623204)              = 0
   ## end of execution of UPX de-compression stub
set_tls(0x127ebc)                       = 0
set_tid_address(0x127e30)               = 731
getuid32()                              = 1000
stat64("/etc/busybox.conf", 0xbea84320) = -1 ENOENT (No such file or directory)
write(2, "foo", 3foo)                      = 3
write(2, ": applet not found\n", 19: applet not found
)    = 19
exit_group(127)                         = ?
+++ exited with 127 +++
```

**Comment by UMRnInside** [comment]
It's not that simple.
Running `upx` in chroot with QEMU produce correct output.
```
apt install cdebootstrap-static qemu-user-static
mkdir foo
cdebootstrap-static -a armel --foreign --include=upx-ucl stable foo/
cp busybox-armv7l foo/
chroot foo/ upx busybox-armv7l
chroot foo/ /busybox armv7l
```

**Comment by jreiser** [comment]
Please give a reproducible recipe.  The `apt install` worked, but:
```
$ sudo cdebootstrap-static -a armel --foreign --include=upx-ucl stable foo/
E: No mirror specified and no default available
```

**Comment by UMRnInside** [comment]
@jreiser Sorry, I missed some stuff
It should be
`
cdebootstrap-static -a armel --foreign --include=upx-ucl stable foo/ http://deb.debian.org/debian/
`

**Comment by jreiser** [comment]
I have reproduced the problem on a RaspberryPi model 2 (32-bit ARM) using `20190628_raspberry-pi-2_buster_PREVIEW.img.xz` which runs `Linux rpi2 4.19.0-5-armmp #1 SMP Debian 4.19.37-5 (2019-06-19) armv7l GNU/Linux`
Please point to the source for the particular busybox-armv7l, or at least a corresponding non-stripped binary.  That will make it easier to find the discrepancy.

**Comment by UMRnInside** [comment]
Oops, there's something **even worse**!
Even "dummy" executable can produce `Invalid argument`
Steps (as root):
```
# Prepare
apt install cdebootstrap-static qemu-user-static gcc-arm-linux-gnueabi upx-ucl
# armel chroot
cdebootstrap-static -a armel --foreign --include=upx-ucl stable foo/ http://deb.debian.org/debian/
# arm dummy executable
echo "int main(int argc, char** argv){ return 0; }" | arm-linux-gnueabi-gcc -o dummy -static -x c -

# upx
upx -o upx-amd64-dummy dummy
cp upx-amd64-dummy dummy armel/
chroot armel/ upx -o upx-armel-dummy dummy
cp armel/upx-armel-dummy .
# Run
./dummy
./upx-amd64-dummy
./upx-armel-dummy
# md5sum
md5sum *dummy
```

The output:
```
root@host:/tmp# echo "int main(int argc, char** argv){ return 0; }" | arm-linux-gnueabi-gcc -o dummy -static -x c -
root@host:/tmp# 
root@host:/tmp# # upx
root@host:/tmp# upx -o upx-amd64-dummy dummy
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2018
UPX 3.95        Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 26th 2018

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
    502220 ->    234576   46.71%    linux/arm    upx-amd64-dummy               

Packed 1 file.
root@host:/tmp# cp upx-amd64-dummy dummy armel/
root@host:/tmp# chroot armel/ upx -o upx-armel-dummy dummy
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2013
UPX 3.91        Markus Oberhumer, Laszlo Molnar & John Reiser   Sep 30th 2013

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
    502220 ->    234428   46.68%   linux/armel   upx-armel-dummy               

Packed 1 file.
root@host:/tmp# cp armel/upx-armel-dummy .
root@host:/tmp# 
root@host:/tmp# ./dummy
root@host:/tmp# ./upx-amd64-dummy
./upx-amd64-dummy: Invalid argument
root@host:/tmp# ./upx-armel-dummy
root@host:/tmp# md5sum *dummy
2907a95b32946ecebb8615d9c9cdf43b  dummy
93330f07c7e398caee86ecee45271af5  upx-amd64-dummy
e46ecfbf5767efd66ddb54562c859e8f  upx-armel-dummy
```

**Comment by UMRnInside** [comment]
After the tests above, I tested `upx -d`

```
upx -d upx-armel-dummy -o upxd-amd64-upx-armel-dummy
chroot armel/ upx -d upx-amd64-dummy -o upxd-armel-upx-amd64-dummy
cp armel/upxd-armel-upx-amd64-dummy .
md5sum *dummy
```

And MD5 checksums:
```
2907a95b32946ecebb8615d9c9cdf43b  dummy
93330f07c7e398caee86ecee45271af5  upx-amd64-dummy
e46ecfbf5767efd66ddb54562c859e8f  upx-armel-dummy
2907a95b32946ecebb8615d9c9cdf43b  upxd-amd64-upx-armel-dummy
2907a95b32946ecebb8615d9c9cdf43b  upxd-armel-upx-amd64-dummy
```

**Comment by jreiser** [comment]
Please upload ("Attach files by ...") the two files `dummy` and `upx-amd64-dummy` that produced this transcript from 2 days ago ( https://github.com/upx/upx/issues/325#issuecomment-569577418 ):
```
root@host:/tmp# ./dummy
root@host:/tmp# ./upx-amd64-dummy
./upx-amd64-dummy: Invalid argument
```
Compress the files with zip, gzip, or .tgz so that github will accept the uploads.

**Comment by UMRnInside** [comment]
@jreiser I made a tar.gz archive of them.
[upx-dummy.tar.gz](https://github.com/upx/upx/files/4013776/upx-dummy.tar.gz)

**Comment by jreiser** [comment]
`upx-armel-dummy` was compressed with an old upx.  Please re-build upx using the tip of the `devel` branch.
```
$ readelf --segments upx-amd64-dummy   ## GNU_STACK present

Elf file type is EXEC (Executable file)
Entry point 0x417d4
There are 3 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x00010000 0x00010000 0x3218f 0x3218f R E 0x10000
  LOAD           0x000000 0x00050000 0x00050000 0x00000 0x379a8 RW  0x10000
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10
$ 
```
```
$ readelf --segments upx-armel-dummy   ## GNU_STACK omitted

Elf file type is EXEC (Executable file)
Entry point 0x417c0
There are 2 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000000 0x00010000 0x00010000 0x320f9 0x320f9 R E 0x10000
  LOAD           0x0079a8 0x000879a8 0x000879a8 0x00000 0x00000 RW  0x10000
$ 
```

**Comment by UMRnInside** [comment]
@jreiser I rebuilt upx using devel branch(on armel). And they are identical now.
```
2907a95b32946ecebb8615d9c9cdf43b  dummy
93330f07c7e398caee86ecee45271af5  upx-amd64-dummy
93330f07c7e398caee86ecee45271af5  upx-armel-dummy
```

**Comment by jreiser** [comment]
As commented above on Jan.04 (21 days ago), the failures occurred in executables that were compressed using an old version of UPX. Using the recently-released upx-3.96, then the compressed executables work as expected.
